File: app/src/main/java/com/music/player/bhandari/m/activity/FragmentArtistInfo.java
package com.music.player.bhandari.m.activity;

import android.annotation.SuppressLint;
import android.content.ActivityNotFoundException;
import android.content.BroadcastReceiver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.MediaStore;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.music.player.bhandari.m.R;
import com.music.player.bhandari.m.model.Constants;
import com.music.player.bhandari.m.model.TrackItem;
import com.music.player.bhandari.m.qlyrics.LyricsAndArtistInfo.ArtistInfo.ArtistInfo;
import com.music.player.bhandari.m.qlyrics.LyricsAndArtistInfo.offlineStorage.OfflineStorageArtistBio;
import com.music.player.bhandari.m.qlyrics.LyricsAndArtistInfo.tasks.DownloadArtInfoThread;
import com.music.player.bhandari.m.interfaces.DoubleClickListener;
import com.music.player.bhandari.m.MyApp;
import com.music.player.bhandari.m.service.PlayerService;
import com.music.player.bhandari.m.utils.UtilityFun;
import com.wang.avi.AVLoadingIndicatorView;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Copyright 2017 Amit Bhandari AB
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class FragmentArtistInfo extends Fragment implements ArtistInfo.Callback {
    private View layout;
    private BroadcastReceiver mArtistUpdateReceiver;
    private ArtistInfo mArtistInfo;
    @BindView(R.id.text_view_art_bio_frag)
    TextView artBioText;
    @BindView(R.id.retry_text_view)
    TextView retryText;
    @BindView(R.id.update_track_metadata)
    TextView updateTagsText;

    @BindView(R.id.loading_lyrics_animation)
    AVLoadingIndicatorView lyricLoadAnimation;
    @BindView(R.id.track_artist_artsi_bio_frag)
    EditText artistEdit;
    @BindView(R.id.button_update_metadata)
    Button buttonUpdateMetadata;

    //@BindView(R.id.ad_view_wrapper) View adViewWrapper;
    //@BindView(R.id.adView)  AdView mAdView;
    //@BindView(R.id.ad_close)  TextView adCloseText;

    private PlayerService playerService;

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        Log.v("frag", isVisibleToUser + "");
        super.setUserVisibleHint(isVisibleToUser);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        layout = inflater.inflate(R.layout.fragment_artist_info, container, false);
        ButterKnife.bind(this, layout);
        playerService = MyApp.getService();

        if (MyApp.getService() == null) {
            UtilityFun.restartApp();
            return layout;
        }
        playerService = MyApp.getService();

        buttonUpdateMetadata.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TrackItem item = playerService.getCurrentTrack();
                if (item == null) {
                    return;
                }

                String edited_artist = artistEdit.getText().toString().trim();

                if (edited_artist.isEmpty()) {
                    Toast.makeText(getContext(), getString(R.string.te_error_empty_field), Toast.LENGTH_SHORT).show();
                    return;
                }


                if (!edited_artist.equals(item.getArtist())) {

                    //changes made, save those
                    Uri uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;

                    ContentValues values = new ContentValues();
                    values.put(MediaStore.Audio.Media.ARTIST, edited_artist);
                    getContext().getContentResolver()
                            .update(uri, values, MediaStore.Audio.Media.TITLE + "=?", new String[]{item.getTitle()});

                    Intent intent = new Intent(getContext(), ActivityNowPlaying.class);
                    intent.putExtra("refresh", true);
                    intent.putExtra("position", playerService.getCurrentTrackPosition());
                    intent.putExtra("originalTitle", item.getTitle());
                    intent.putExtra("title", item.getTitle());
                    intent.putExtra("artist", edited_artist);
                    intent.putExtra("album", item.getAlbum());
                    startActivity(intent);

                    artistEdit.setVisibility(View.GONE);
                    updateTagsText.setVisibility(View.GONE);
                    buttonUpdateMetadata.setVisibility(View.GONE);
                    buttonUpdateMetadata.setClickable(false);


                    if (getActivity() != null) {
                        View view = getActivity().getCurrentFocus();
                        if (view != null) {
                            InputMethodManager imm = (InputMethodManager) getActivity()
                                    .getSystemService(Context.INPUT_METHOD_SERVICE);
                            if (imm != null) {
                                imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
                            }
                        }
                    }

                    downloadArtInfo();

                } else {
                    Toast.makeText(getContext(), getString(R.string.change_tags_to_update), Toast.LENGTH_SHORT).show();
                }
            }
        });

        //retry click listner
        layout.findViewById(R.id.ll_art_bio).setOnClickListener(new DoubleClickListener() {
            @Override
            public void onSingleClick(View v) {
                if (retryText.getVisibility() == View.VISIBLE) {
                    retryText.setVisibility(View.GONE);
                    artBioText.setVisibility(View.VISIBLE);
                    artistEdit.setVisibility(View.GONE);
                    updateTagsText.setVisibility(View.GONE);
                    buttonUpdateMetadata.setVisibility(View.GONE);
                    buttonUpdateMetadata.setClickable(false);
                    lyricLoadAnimation.setVisibility(View.GONE);
                    downloadArtInfo();
                }
            }

            @Override
            public void onDoubleClick(View v) {

                //if no connection text, do not hide artist content
                if (retryText.getText().toString().equals(getString(R.string.no_connection))) {
                    return;
                }

                if (artBioText.getVisibility() == View.VISIBLE) {
                    artBioText.setVisibility(View.GONE);
                } else {
                    artBioText.setVisibility(View.VISIBLE);
                }
            }
        });

        //downloadArtInfo();
        mArtistUpdateReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                //already displayed, skip
                updateArtistInfoIfNeeded();
            }
        };

        return layout;
    }

    private void downloadArtInfo() {
        TrackItem item = playerService.getCurrentTrack();
        if (item == null || item.getArtist() == null) {
            return;
        }

        artBioText.setText(getString(R.string.artist_info_loading));

        //set loading animation
        lyricLoadAnimation.setVisibility(View.VISIBLE);
        lyricLoadAnimation.show();

        //see in offlinne db first
        mArtistInfo = OfflineStorageArtistBio.getArtistBioFromTrackItem(item);
        //second check is added to make sure internet call will happen
        //when user manually changes artist tag
        if (mArtistInfo != null && item.getArtist().trim().equals(mArtistInfo.getOriginalArtist().trim())) {
            onArtInfoDownloaded(mArtistInfo);
            return;
        }

        if (UtilityFun.isConnectedToInternet()) {

            String artist = item.getArtist();
            artist = UtilityFun.filterArtistString(artist);

            new DownloadArtInfoThread(this, artist, item).start();
        } else {
            artBioText.setVisibility(View.GONE);
            retryText.setText(getString(R.string.no_connection));
            retryText.setVisibility(View.VISIBLE);
            lyricLoadAnimation.hide();
            lyricLoadAnimation.setVisibility(View.GONE);
        }

    }

    @Override
    public void onPause() {
        super.onPause();
        LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(mArtistUpdateReceiver);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (MyApp.getService() != null) {
            updateArtistInfoIfNeeded();
            LocalBroadcastManager.getInstance(getContext()).registerReceiver(mArtistUpdateReceiver
                    , new IntentFilter(Constants.ACTION.UPDATE_LYRIC_AND_INFO));
        } else {
            UtilityFun.restartApp();
        }
    }

    private void updateArtistInfoIfNeeded() {
        TrackItem item = playerService.getCurrentTrack();
        if (item == null) {
            artBioText.setVisibility(View.GONE);
            retryText.setText(getString(R.string.no_music_found));
            //retryText.setVisibility(View.GONE);
            retryText.setVisibility(View.VISIBLE);
            lyricLoadAnimation.hide();
            return;
        }
        if (mArtistInfo != null && mArtistInfo.getOriginalArtist()
                .equals(item.getArtist())) {
            return;
        }

        //set loading  text and animation
        //set loading  text and animation

        downloadArtInfo();
    }

    @Override
    public void onArtInfoDownloaded(ArtistInfo artistInfo) {

        mArtistInfo = artistInfo;
        if (artistInfo == null || getActivity() == null || !isAdded()) {
            return;
        }
        TrackItem item = playerService.getCurrentTrack();
        //if song is already changed , return
        if (item != null && !item.getArtist().trim().equals(artistInfo.getOriginalArtist().trim())) {
            //artBioText.setText(getString(R.string.artist_info_loading));
            return;
        }
        //hide loading animation
        lyricLoadAnimation.hide();
        lyricLoadAnimation.setVisibility(View.GONE);

        if (artistInfo.getArtistContent() == null) {
            retryText.setText(getString(R.string.artist_info_no_result));
            retryText.setVisibility(View.VISIBLE);
            artBioText.setVisibility(View.GONE);
            TrackItem tempItem = playerService.getCurrentTrack();
            if (tempItem != null) {
                artistEdit.setVisibility(View.VISIBLE);
                updateTagsText.setVisibility(View.VISIBLE);
                buttonUpdateMetadata.setVisibility(View.VISIBLE);
                buttonUpdateMetadata.setClickable(true);
                artistEdit.setText(tempItem.getArtist());
            }
            return;
        }

        if (layout != null && getActivity() != null && artistInfo.getArtistContent() != null) {
            Log.d("onArtInfoDownloaded", "onArtInfoDownloaded: " + artistInfo.getCorrectedArtist());
            String content = artistInfo.getArtistContent();
            int index = content.indexOf("Read more");
            SpannableString ss = new SpannableString(content);
            ClickableSpan clickableSpan = new ClickableSpan() {
                @Override
                public void onClick(View textView) {
                    if (mArtistInfo.getArtistUrl() == null) {
                        Toast.makeText(getContext(), getString(R.string.error_invalid_url), Toast.LENGTH_SHORT).show();
                    } else {
                        try {
                            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(mArtistInfo.getArtistUrl()));
                            startActivity(browserIntent);
                        } catch (ActivityNotFoundException e) {
                            Toast.makeText(getContext(), "No supporting application found for opening the link.", Toast.LENGTH_SHORT).show();
                        }
                    }
                }

                @Override
                public void updateDrawState(TextPaint ds) {
                    super.updateDrawState(ds);
                    ds.setUnderlineText(true);
                    ds.setTypeface(Typeface.create(ds.getTypeface(), Typeface.BOLD));
                }
            };
            if (index != -1) {
                ss.setSpan(clickableSpan, index, index + 9, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            }


            if (!content.equals("")) {
                artBioText.setVisibility(View.VISIBLE);
                retryText.setVisibility(View.GONE);

                artBioText.setText(ss);
                artBioText.setMovementMethod(LinkMovementMethod.getInstance());

                artistEdit.setVisibility(View.GONE);
                updateTagsText.setVisibility(View.GONE);
                buttonUpdateMetadata.setVisibility(View.GONE);
                buttonUpdateMetadata.setClickable(false);
                artistEdit.setText("");

            } else {
                artBioText.setVisibility(View.GONE);
                retryText.setText(getString(R.string.artist_info_no_result));
                retryText.setVisibility(View.VISIBLE);
                TrackItem tempItem = playerService.getCurrentTrack();
                if (tempItem != null) {
                    artistEdit.setVisibility(View.VISIBLE);
                    updateTagsText.setVisibility(View.VISIBLE);
                    buttonUpdateMetadata.setVisibility(View.VISIBLE);
                    buttonUpdateMetadata.setClickable(true);
                    artistEdit.setText(tempItem.getArtist());
                }
            }

            //check current now playing background setting
            ///get current setting
            // 0 - System default   1 - artist image  2 - custom
            int currentNowPlayingBackPref = MyApp.getPref().getInt(getString(R.string.pref_now_playing_back), 1);

            if (currentNowPlayingBackPref == 1 && !artistInfo.getCorrectedArtist().equals("[unknown]")) {
                if (!((ActivityNowPlaying) getActivity()).isArtistLoadedInBack()) {
                    new SetBlurryImagetask().execute(artistInfo);
                }
            }
        }
    }

    @SuppressLint("StaticFieldLeak")
    private class SetBlurryImagetask extends AsyncTask<ArtistInfo, String, Bitmap> {

        Bitmap b;

        @Override
        protected Bitmap doInBackground(ArtistInfo... params) {

            //store file in cache with artist id as name
            //create folder in cache for artist images
            String CACHE_ART_THUMBS = MyApp.getContext().getCacheDir() + "/art_thumbs/";
            String actual_file_path = CACHE_ART_THUMBS + params[0].getOriginalArtist();
            File f = new File(CACHE_ART_THUMBS);
            if (!f.exists()) {
                f.mkdir();
            }
            if (!new File(actual_file_path).exists()) {
                //create file
                FileOutputStream fos = null;
                try {
                    fos = new FileOutputStream(new File(actual_file_path));
                    URL url = new URL(params[0].getImageUrl());
                    InputStream inputStream = url.openConnection().getInputStream();
                    byte[] buffer = new byte[1024];
                    int bufferLength = 0;
                    while ((bufferLength = inputStream.read(buffer)) > 0) {
                        fos.write(buffer, 0, bufferLength);
                    }
                    fos.close();
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }

            }

            b = BitmapFactory.decodeFile(actual_file_path);
            return b;
        }

        protected void onPostExecute(Bitmap b) {

            //set background image

            if (b != null && getActivity() != null) {
                ((ActivityNowPlaying) getActivity()).setBlurryBackground(b);
            }
        }
    }
}

File: app/src/main/java/com/music/player/bhandari/m/activity/FragmentArtistInfo.kt
package com.music.player.bhandari.m.activity

import android.annotation.SuppressLint
import android.content.*
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Typeface
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.provider.MediaStore
import android.text.SpannableString
import android.text.Spanned
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import butterknife.BindView
import butterknife.ButterKnife
import com.music.player.bhandari.m.MyApp
import com.music.player.bhandari.m.R
import com.music.player.bhandari.m.interfaces.DoubleClickListener
import com.music.player.bhandari.m.model.Constants
import com.music.player.bhandari.m.model.TrackItem
import com.music.player.bhandari.m.qlyrics.LyricsAndArtistInfo.ArtistInfo.ArtistInfo
import com.music.player.bhandari.m.qlyrics.LyricsAndArtistInfo.offlineStorage.OfflineStorageArtistBio
import com.music.player.bhandari.m.qlyrics.LyricsAndArtistInfo.tasks.DownloadArtInfoThread
import com.music.player.bhandari.m.service.PlayerService
import com.music.player.bhandari.m.utils.UtilityFun
import com.wang.avi.AVLoadingIndicatorView
import java.io.*
import java.net.URL

/**
 * Copyright 2017 Amit Bhandari AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class FragmentArtistInfo : Fragment(), ArtistInfo.Callback {
    private var layout: View? = null
    private var mArtistUpdateReceiver: BroadcastReceiver? = null
    private var mArtistInfo: ArtistInfo? = null

    @BindView(R.id.text_view_art_bio_frag)
    var artBioText: TextView? = null

    @BindView(R.id.retry_text_view)
    var retryText: TextView? = null

    @BindView(R.id.update_track_metadata)
    var updateTagsText: TextView? = null

    @BindView(R.id.loading_lyrics_animation)
    var lyricLoadAnimation: AVLoadingIndicatorView? = null

    @BindView(R.id.track_artist_artsi_bio_frag)
    var artistEdit: EditText? = null

    @BindView(R.id.button_update_metadata)
    var buttonUpdateMetadata: Button? = null

    //@BindView(R.id.ad_view_wrapper) View adViewWrapper;
    //@BindView(R.id.adView)  AdView mAdView;
    //@BindView(R.id.ad_close)  TextView adCloseText;
    private var playerService: PlayerService? = null
    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        Log.v("frag", isVisibleToUser.toString() + "")
        /*if(isVisibleToUser && mAdView!=null){
            mAdView.resume();
        }else {
            if(mAdView!=null){
                mAdView.pause();
            }
        }*/super.setUserVisibleHint(isVisibleToUser)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        layout = inflater.inflate(R.layout.fragment_artist_info, container, false)
        ButterKnife.bind(this, layout!!)
        playerService = MyApp.getService()
        if (MyApp.getService() == null) {
            UtilityFun.restartApp()
            return layout
        }
        playerService = MyApp.getService()
        buttonUpdateMetadata!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View) {
                val item: TrackItem = playerService!!.getCurrentTrack() ?: return
                val edited_artist: String = artistEdit!!.text.toString().trim { it <= ' ' }
                if (edited_artist.isEmpty()) {
                    Toast.makeText(context,
                        getString(R.string.te_error_empty_field),
                        Toast.LENGTH_SHORT).show()
                    return
                }
                if (!(edited_artist == item.getArtist())) {

                    //changes made, save those
                    val uri: Uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
                    val values: ContentValues = ContentValues()
                    values.put(MediaStore.Audio.Media.ARTIST, edited_artist)
                    context!!.contentResolver
                        .update(uri,
                            values,
                            MediaStore.Audio.Media.TITLE + "=?",
                            arrayOf(item.title))
                    val intent: Intent = Intent(context, ActivityNowPlaying::class.java)
                    intent.putExtra("refresh", true)
                    intent.putExtra("position", playerService!!.getCurrentTrackPosition())
                    intent.putExtra("originalTitle", item.title)
                    intent.putExtra("title", item.title)
                    intent.putExtra("artist", edited_artist)
                    intent.putExtra("album", item.album)
                    startActivity(intent)
                    artistEdit!!.visibility = View.GONE
                    updateTagsText!!.visibility = View.GONE
                    buttonUpdateMetadata!!.visibility = View.GONE
                    buttonUpdateMetadata!!.isClickable = false
                    if (activity != null) {
                        val view: View? = activity!!.currentFocus
                        if (view != null) {
                            (activity!!.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?)?.hideSoftInputFromWindow(
                                view.windowToken,
                                0)
                        }
                    }
                    downloadArtInfo()
                } else {
                    Toast.makeText(context,
                        getString(R.string.change_tags_to_update),
                        Toast.LENGTH_SHORT).show()
                }
            }
        })

        //retry click listner
        layout!!.findViewById<View>(R.id.ll_art_bio)
            .setOnClickListener(object : DoubleClickListener() {
                override fun onSingleClick(v: View?) {
                    if (retryText!!.visibility == View.VISIBLE) {
                        retryText!!.visibility = View.GONE
                        artBioText!!.visibility = View.VISIBLE
                        artistEdit!!.visibility = View.GONE
                        updateTagsText!!.visibility = View.GONE
                        buttonUpdateMetadata!!.visibility = View.GONE
                        buttonUpdateMetadata!!.isClickable = false
                        lyricLoadAnimation!!.visibility = View.GONE
                        downloadArtInfo()
                    }
                }

                override fun onDoubleClick(v: View?) {

                    //if no connection text, do not hide artist content
                    if ((retryText!!.text.toString() == getString(R.string.no_connection))) {
                        return
                    }
                    if (artBioText!!.visibility == View.VISIBLE) {
                        artBioText!!.visibility = View.GONE
                    } else {
                        artBioText!!.visibility = View.VISIBLE
                    }
                }
            })

        //downloadArtInfo();
        mArtistUpdateReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                //already displayed, skip
                updateArtistInfoIfNeeded()
            }
        }
        return layout
    }

    private fun downloadArtInfo() {
        val item: TrackItem? = playerService!!.getCurrentTrack()
        if (item?.getArtist() == null) {
            return
        }
        artBioText!!.setText(getString(R.string.artist_info_loading))

        //set loading animation
        lyricLoadAnimation!!.visibility = View.VISIBLE
        lyricLoadAnimation!!.show()

        //see in offlinne db first
        mArtistInfo = OfflineStorageArtistBio.getArtistBioFromTrackItem(item)
        //second check is added to make sure internet call will happen
        //when user manually changes artist tag
        if (mArtistInfo != null && (item.getArtist()!!
                .trim { it <= ' ' } == mArtistInfo!!.getOriginalArtist()!!.trim())
        ) {
            onArtInfoDownloaded(mArtistInfo)
            return
        }
        if (UtilityFun.isConnectedToInternet) {
            var artist: String? = item.getArtist()
            artist = UtilityFun.filterArtistString(artist!!)
            DownloadArtInfoThread(this, artist, item).start()
        } else {
            artBioText!!.visibility = View.GONE
            retryText!!.text = getString(R.string.no_connection)
            retryText!!.visibility = View.VISIBLE
            lyricLoadAnimation!!.hide()
            lyricLoadAnimation!!.visibility = View.GONE
        }
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(mArtistUpdateReceiver!!)
    }

    override fun onResume() {
        super.onResume()
        if (MyApp.getService() != null) {
            updateArtistInfoIfNeeded()
            LocalBroadcastManager.getInstance(requireContext()).registerReceiver(mArtistUpdateReceiver!!,
                IntentFilter(Constants.ACTION.UPDATE_LYRIC_AND_INFO))
        } else {
            UtilityFun.restartApp()
        }
    }

    private fun updateArtistInfoIfNeeded() {
        val item: TrackItem? = playerService!!.getCurrentTrack()
        if (item == null) {
            artBioText!!.visibility = View.GONE
            retryText!!.text = getString(R.string.no_music_found)
            //retryText.setVisibility(View.GONE);
            retryText!!.visibility = View.VISIBLE
            lyricLoadAnimation!!.hide()
            return
        }
        if (mArtistInfo != null && mArtistInfo!!.getOriginalArtist().equals(item.getArtist())) {
            return
        }

        //set loading  text and animation
        //set loading  text and animation
        downloadArtInfo()
    }

    override fun onArtInfoDownloaded(artistInfo: ArtistInfo?) {
        mArtistInfo = artistInfo
        if ((artistInfo == null) || (activity == null) || !isAdded) {
            return
        }
        val item: TrackItem? = playerService!!.getCurrentTrack()
        //if song is already changed , return
        if (item != null && !(item.getArtist()!!
                .trim { it <= ' ' } == artistInfo.getOriginalArtist()!!.trim())
        ) {
            //artBioText.setText(getString(R.string.artist_info_loading));
            return
        }
        //hide loading animation
        lyricLoadAnimation!!.hide()
        lyricLoadAnimation!!.visibility = View.GONE
        if (artistInfo.getArtistContent() == null) {
            retryText!!.text = getString(R.string.artist_info_no_result)
            retryText!!.visibility = View.VISIBLE
            artBioText!!.visibility = View.GONE
            val tempItem: TrackItem? = playerService!!.getCurrentTrack()
            if (tempItem != null) {
                artistEdit!!.visibility = View.VISIBLE
                updateTagsText!!.visibility = View.VISIBLE
                buttonUpdateMetadata!!.visibility = View.VISIBLE
                buttonUpdateMetadata!!.isClickable = true
                artistEdit!!.setText(tempItem.getArtist())
            }
            return
        }
        if ((layout != null) && (activity != null) && (artistInfo.getArtistContent() != null)) {
            Log.d("onArtInfoDownloaded", "onArtInfoDownloaded: " + artistInfo.getCorrectedArtist())
            val content = artistInfo.getArtistContent()
            val index: Int = content!!.indexOf("Read more")
            val ss = SpannableString(content)
            val clickableSpan: ClickableSpan = object : ClickableSpan() {
                override fun onClick(textView: View) {
                    if (mArtistInfo!!.getArtistUrl() == null) {
                        Toast.makeText(context,
                            getString(R.string.error_invalid_url),
                            Toast.LENGTH_SHORT).show()
                    } else {
                        try {
                            val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(
                                mArtistInfo!!.getArtistUrl()))
                            startActivity(browserIntent)
                        } catch (e: ActivityNotFoundException) {
                            Toast.makeText(context,
                                "No supporting application found for opening the link.",
                                Toast.LENGTH_SHORT).show()
                        }
                    }
                }

                override fun updateDrawState(ds: TextPaint) {
                    super.updateDrawState(ds)
                    ds.isUnderlineText = true
                    ds.typeface = Typeface.create(ds.typeface, Typeface.BOLD)
                }
            }
            if (index != -1) {
                ss.setSpan(clickableSpan, index, index + 9, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            if (!(content == "")) {
                artBioText!!.visibility = View.VISIBLE
                retryText!!.visibility = View.GONE
                artBioText!!.text = ss
                artBioText!!.movementMethod = LinkMovementMethod.getInstance()
                artistEdit!!.visibility = View.GONE
                updateTagsText!!.visibility = View.GONE
                buttonUpdateMetadata!!.visibility = View.GONE
                buttonUpdateMetadata!!.isClickable = false
                artistEdit!!.setText("")
            } else {
                artBioText!!.visibility = View.GONE
                retryText!!.text = getString(R.string.artist_info_no_result)
                retryText!!.visibility = View.VISIBLE
                val tempItem: TrackItem? = playerService!!.getCurrentTrack()
                if (tempItem != null) {
                    artistEdit!!.visibility = View.VISIBLE
                    updateTagsText!!.visibility = View.VISIBLE
                    buttonUpdateMetadata!!.visibility = View.VISIBLE
                    buttonUpdateMetadata!!.isClickable = true
                    artistEdit!!.setText(tempItem.getArtist())
                }
            }

            //check current now playing background setting
            ///get current setting
            // 0 - System default   1 - artist image  2 - custom
            val currentNowPlayingBackPref: Int =
                MyApp.getPref()!!.getInt(getString(R.string.pref_now_playing_back), 1)
            if (currentNowPlayingBackPref == 1 && !artistInfo.getCorrectedArtist()
                    .equals("[unknown]")
            ) {
                if (!(activity as ActivityNowPlaying?)!!.isArtistLoadedInBack()) {
                    SetBlurryImagetask().execute(artistInfo)
                }
            }
        }
    }

    @SuppressLint("StaticFieldLeak")
    private inner class SetBlurryImagetask :
        AsyncTask<ArtistInfo?, String?, Bitmap?>() {
        var b: Bitmap? = null
        override fun doInBackground(vararg params: ArtistInfo?): Bitmap? {

            //store file in cache with artist id as name
            //create folder in cache for artist images
            val CACHE_ART_THUMBS: String =
                MyApp.getContext()!!.cacheDir.toString() + "/art_thumbs/"
            val actual_file_path: String = CACHE_ART_THUMBS + params[0]!!.getOriginalArtist()
            val f = File(CACHE_ART_THUMBS)
            if (!f.exists()) {
                f.mkdir()
            }
            if (!File(actual_file_path).exists()) {
                //create file
                val fos: FileOutputStream?
                try {
                    fos = FileOutputStream(File(actual_file_path))
                    val url = URL(params[0]!!.getImageUrl())
                    val inputStream: InputStream = url.openConnection().getInputStream()
                    val buffer = ByteArray(1024)
                    var bufferLength: Int
                    while ((inputStream.read(buffer).also { bufferLength = it }) > 0) {
                        fos.write(buffer, 0, bufferLength)
                    }
                    fos.close()
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            b = BitmapFactory.decodeFile(actual_file_path)
            return b
        }

        override fun onPostExecute(b: Bitmap?) {

            //set background image
            if (b != null && activity != null) {
                (activity as ActivityNowPlaying?)!!.setBlurryBackground(b)
            }
        }
    }
}